{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import nbinom\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "from scipy.stats import norm\n",
    "from IPython.display import display, HTML\n",
    "from matplotlib.pyplot import figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "# This script will generate a simulated read count file for a disomy and trisomy dataset that can be fed directly into\n",
    "# DESEQ2. Essentially, a previous RNA-seq dataset is used as the basis for finding mu for each gene. Replicates from \n",
    "# Eric are used to sample first from a normal distribution to generate mu, and then the\n",
    "# dispersion is calculated through parametric fit (y = a + b/x) where a is asymptotic dispersion and b is extra-poisson\n",
    "# noise. a and b thus act as hyperparameters for our sampling. Additional parameters include the number of replicates in\n",
    "# a single experiment, the number of experiments (trials), and a scaling factor to increase the depth of the samples\n",
    "\n",
    "\n",
    "trials=3 # The number of experiments to generate. A new file will be generated with the same parameters for [1:trials)\n",
    "\n",
    "disomy_disp=.01 # The asymptotic dispersion (hyperparameter a) for the disomy simulations\n",
    "disomy_extrapoiss=1 # The extra-poisson noise (hyperparameter b) for the disomy simulations\n",
    "asymptDisp=disomy_disp\n",
    "extraPois=disomy_extrapoiss\n",
    "\n",
    "test_array=np.linspace(3,10,8).astype(int) # numpy list for iterating over multiple values of the hyperparameters or replicate/scales\n",
    "\n",
    "for test_num in test_array: # I just use this to iterate through values I'm interested in testing (dispersion, poisson noise, depth, or replication)\n",
    "    print(test_num)\n",
    "    trisomy_disp=.03 # The asymptotic dispersion for trisomy simulation (a)\n",
    "    trisomy_extrapoiss=5 # The extrapoisson noise for trisomy simulation (b)\n",
    "    scaling=3 # a scalar to increase depth. Each value of the original mean counts will be multiplied by this to simulate more depth\n",
    "    replicate_num=test_num # How many replicates to make for a single experiment (e.g., 3 will generate 3 disomic and 3 trisomic datasets)\n",
    "    for trial_num in range(1,trials):\n",
    "        # DISOMY SECTION\n",
    "        # We first read in a real count file generated from biological data\n",
    "        counts_df = pd.read_csv('/Shares/down/mixed/RNAGRO01132021/data/RNAseqfiles/res_featureCounts_gene_idfull_143138.coverage.csv')\n",
    "        # First, we'll generate a normal distribution for the mean\n",
    "        # We'll calculate a mean and variance for each gene from Eric, and sample from that to get our simulated mean\n",
    "        \n",
    "        # We also need an annotation df to assess which genes belong to chr21\n",
    "        annotation_df = pd.read_csv('/Shares/down/mixed/RNAGRO01132021/data/RNAseqfiles/res_featureCounts_gene_idfull_143138.annotation.csv')\n",
    "        annotation_df=annotation_df[annotation_df['Chr'].str.match('chr21')]\n",
    "        chr21genes = annotation_df['GeneID']       \n",
    "        \n",
    "        # First we average the counts from Eric's columns, and find the variance\n",
    "        erics_cols = ['Eric_repA_tech1.sorted.bam', 'Eric_repB_tech1.2.sorted.bam','Eric_repC_tech1.sorted.bam']\n",
    "        counts_df['Eric_average'] = np.ceil(counts_df[erics_cols].mean(axis=1)) \n",
    "        counts_df['Eric_variance'] = counts_df[erics_cols].var(axis=1) \n",
    "        genenames= counts_df['Unnamed: 0']\n",
    "        basecounts= counts_df['Eric_repA_tech1.sorted.bam']\n",
    "        d = {'Gene_id':genenames,'Eric_counts':basecounts}\n",
    "        simulated_df = pd.DataFrame(d) # This dataframe is what we'll actually be adding to for each replicate\n",
    "\n",
    "        means=counts_df['Eric_average'] # This is used as input for sampling from a normal distribution\n",
    "        variances=counts_df['Eric_variance'] # Same for this one\n",
    "        counts_df['diploidy'] = np.where( # Disomy individuals have the same ploidy for each gene, but I include it as a np.where if you wanted to play with this a bit\n",
    "            counts_df['Unnamed: 0'].isin(chr21genes),                # Condition (disomy).\n",
    "            1,  # . \n",
    "            1     # What to assign to price when condition is false. \n",
    "        )\n",
    "        # Here are the sims for the D21 individual. We need to modify our numbers for the T21 individual\n",
    "        sim_name=\"Elvis\"# Just a silly name to save our disomy file to\n",
    "        \n",
    "        # We'll just assign p and n for means at or below zero to avoid infinite dispersion\n",
    "        # Same for samples with 0 variance to prevent infinite p\n",
    "        # Note that since mean ~ variance in NB, this will only occur with means of 0 anyway, so these'll be discarded when\n",
    "        # we run DESeq2 \n",
    "        for j in range(replicate_num):    \n",
    "            counts_df['Eric_disomy'] = (counts_df['Eric_average'] * counts_df['diploidy'])\n",
    "            means=counts_df['Eric_disomy']\n",
    "            simulated_counts=[]\n",
    "            for i in range(len(means)):\n",
    "                #newmean=means[i] #Uncomment this if you don't want to sample from a normal first\n",
    "                newmean=(np.ceil(norm.rvs(size=1,loc=((means[i]+1)*scaling),scale=math.sqrt(variances[i]))))\n",
    "                if newmean <= 0:\n",
    "                    newmean=0 # This prevents negative dispersions, as counts should never go below 0\n",
    "                    dispersion=asymptDisp+(extraPois/1)\n",
    "                else:\n",
    "                    dispersion=asymptDisp+(extraPois/newmean)            \n",
    "                var=newmean+(dispersion*(newmean**2))\n",
    "                if var == 0:\n",
    "                    var=1 # This prevents dividing by 0, but won't change the value of p or n\n",
    "                p=newmean/var\n",
    "                n=(newmean**2)/((var)-newmean)\n",
    "                if (p == 0) and (n==0): # We can just assign r as 0 to avoid undefined values in p and n are 0. This only happens if the original mean is 0\n",
    "                    r = 0\n",
    "                    simulated_counts.append(r)\n",
    "                    continue   \n",
    "                r = nbinom.rvs(n, p, size=1) # sample from negative binomial with parameters n and p\n",
    "                simulated_counts.append(r[0])\n",
    "            colname=sim_name+\"_rep\"+str(j+1)+\"_counts\"\n",
    "            simulated_df[colname]=simulated_counts \n",
    "        sim_name=\"EDad\"# Just a silly name to save our disomy file to\n",
    "        \n",
    "        # We'll just assign p and n for means at or below zero to avoid infinite dispersion\n",
    "        # Same for samples with 0 variance to prevent infinite p\n",
    "        # Note that since mean ~ variance in NB, this will only occur with means of 0 anyway, so these'll be discarded when\n",
    "        # we run DESeq2 \n",
    "        for j in range(replicate_num):    \n",
    "            counts_df['Eric_disomy'] = (counts_df['Eric_average'] * counts_df['diploidy'])\n",
    "            means=counts_df['Eric_disomy']\n",
    "            simulated_counts=[]\n",
    "            for i in range(len(means)):\n",
    "                #newmean=means[i] #Uncomment this if you don't want to sample from a normal first\n",
    "                newmean=(np.ceil(norm.rvs(size=1,loc=((means[i]+1)*scaling),scale=math.sqrt(variances[i]))))\n",
    "                if newmean <= 0:\n",
    "                    newmean=0 # This prevents negative dispersions, as counts should never go below 0\n",
    "                    dispersion=asymptDisp+(extraPois/1)\n",
    "                else:\n",
    "                    dispersion=asymptDisp+(extraPois/newmean)            \n",
    "                var=newmean+(dispersion*(newmean**2))\n",
    "                if var == 0:\n",
    "                    var=1 # This prevents dividing by 0, but won't change the value of p or n\n",
    "                p=newmean/var\n",
    "                n=(newmean**2)/((var)-newmean)\n",
    "                if (p == 0) and (n==0): # We can just assign r as 0 to avoid undefined values in p and n are 0. This only happens if the original mean is 0\n",
    "                    r = 0\n",
    "                    simulated_counts.append(r)\n",
    "                    continue   \n",
    "                r = nbinom.rvs(n, p, size=1) # sample from negative binomial with parameters n and p\n",
    "                simulated_counts.append(r[0])\n",
    "            colname=sim_name+\"_rep\"+str(j+1)+\"_counts\"\n",
    "            simulated_df[colname]=simulated_counts \n",
    "\n",
    "        sim_name=\"EMom\"# Just a silly name to save our disomy file to\n",
    "        \n",
    "        # We'll just assign p and n for means at or below zero to avoid infinite dispersion\n",
    "        # Same for samples with 0 variance to prevent infinite p\n",
    "        # Note that since mean ~ variance in NB, this will only occur with means of 0 anyway, so these'll be discarded when\n",
    "        # we run DESeq2 \n",
    "        for j in range(replicate_num):    \n",
    "            counts_df['Eric_disomy'] = (counts_df['Eric_average'] * counts_df['diploidy'])\n",
    "            means=counts_df['Eric_disomy']\n",
    "            simulated_counts=[]\n",
    "            for i in range(len(means)):\n",
    "                #newmean=means[i] #Uncomment this if you don't want to sample from a normal first\n",
    "                newmean=(np.ceil(norm.rvs(size=1,loc=((means[i]+1)*scaling),scale=math.sqrt(variances[i]))))\n",
    "                if newmean <= 0:\n",
    "                    newmean=0 # This prevents negative dispersions, as counts should never go below 0\n",
    "                    dispersion=asymptDisp+(extraPois/1)\n",
    "                else:\n",
    "                    dispersion=asymptDisp+(extraPois/newmean)            \n",
    "                var=newmean+(dispersion*(newmean**2))\n",
    "                if var == 0:\n",
    "                    var=1 # This prevents dividing by 0, but won't change the value of p or ns\n",
    "                p=newmean/var\n",
    "                n=(newmean**2)/((var)-newmean)\n",
    "                if (p == 0) and (n==0): # We can just assign r as 0 to avoid undefined values in p and n are 0. This only happens if the original mean is 0\n",
    "                    r = 0\n",
    "                    simulated_counts.append(r)\n",
    "                    continue   \n",
    "                r = nbinom.rvs(n, p, size=1) # sample from negative binomial with parameters n and p\n",
    "                simulated_counts.append(r[0])\n",
    "            colname=sim_name+\"_rep\"+str(j+1)+\"_counts\"\n",
    "            simulated_df[colname]=simulated_counts \n",
    "            \n",
    "        # This calculates a sum which you should check against DESeq2's size factor normalization if you want to make sure \n",
    "        column_number=len(simulated_df.columns)\n",
    "        norm_sum=simulated_df.iloc[:, column_number-replicate_num:column_number].sum().mean()\n",
    "        \n",
    "        \n",
    "        # TRISOMY SECTION\n",
    "        annotation_df = pd.read_csv('/Shares/down/mixed/RNAGRO01132021/data/RNAseqfiles/res_featureCounts_gene_idfull_143138.annotation.csv')\n",
    "        annotation_df=annotation_df[annotation_df['Chr'].str.match('chr21')]\n",
    "        counts_df = pd.read_csv('/Shares/down/mixed/RNAGRO01132021/data/RNAseqfiles/res_featureCounts_gene_idfull_143138.coverage.csv')\n",
    "\n",
    "        counts_df['Eric_average'] = np.ceil(counts_df[erics_cols].mean(axis=1)) \n",
    "        counts_df['Eric_variance'] = counts_df[erics_cols].var(axis=1)\n",
    "\n",
    "        chr21genes = annotation_df['GeneID']\n",
    "        counts_df['ploidy'] = np.where(\n",
    "            counts_df['Unnamed: 0'].isin(chr21genes),                # Condition (belongs to chr21).\n",
    "            1.1,  # What to assign to price when condition is true. \n",
    "            1     # What to assign to price when condition is false.\n",
    "        )\n",
    "        gene_ploidy=counts_df['ploidy']\n",
    "\n",
    "        # Here are the sims for the T21 individual. We need to modify our numbers for the T21 individual\n",
    "        sim_name=\"Eddie\" # Silly name for T21 sim. Change as you like!\n",
    "\n",
    "        for j in range(replicate_num):\n",
    "            counts_df['Eric_trisomy'] = (counts_df['Eric_average'] * counts_df['ploidy'])\n",
    "            means=counts_df['Eric_trisomy']\n",
    "            variances=counts_df['Eric_variance']\n",
    "            simulated_counts=[]\n",
    "            for i in range(len(means)):\n",
    "                if gene_ploidy[i]==1.1: # If the gene is triploid, we'll assign the trisomy dispersion\n",
    "                    asymptDisp=trisomy_disp\n",
    "                    extraPois=trisomy_extrapoiss\n",
    "                else: # Otherwise we assign the original values\n",
    "                    asymptDisp=disomy_disp\n",
    "                    extraPois=disomy_extrapoiss\n",
    "                # Note that this is NOT the way it works in DESeq2, where all gene dispersions are used to inform\n",
    "                # the parametric fit. We do it here to demonstrate the effect of trisomy genes having higher variance\n",
    "                # within an individual sample\n",
    "                newmean=(np.ceil(norm.rvs(size=1,loc=((means[i]+1)*scaling),scale=math.sqrt(variances[i])))) # Scale up mean as desired to represent more depth\n",
    "                if newmean <= 0:\n",
    "                    newmean=0\n",
    "                    dispersion=asymptDisp+(extraPois/1)\n",
    "                else:\n",
    "                    dispersion=asymptDisp+(extraPois/newmean)\n",
    "                var=newmean+(dispersion*(newmean**2))\n",
    "                if var == 0:\n",
    "                    var=1\n",
    "                p=newmean/var\n",
    "                n=(newmean**2)/((var)-newmean)\n",
    "                if (p == 0) and (n==0):\n",
    "                    r = 0\n",
    "                    simulated_counts.append(r)\n",
    "                    continue   \n",
    "                r = nbinom.rvs(n, p, size=1)\n",
    "                simulated_counts.append(r[0])\n",
    "            colname=sim_name+\"_rep\"+str(j+1)+\"_counts\"\n",
    "            #norm_factor=(norm_sum/sum(simulated_counts)) # You can re-normalize the counts to be the same depth by uncommenting this\n",
    "            norm_factor=1 # And commenting this instead\n",
    "            simulated_counts = [np.ceil(norm_factor*x) for x in simulated_counts]\n",
    "            simulated_df[colname]=simulated_counts  \n",
    "\n",
    "\n",
    "        simulated_df = simulated_df.drop(columns=['Eric_counts'])\n",
    "        filename=\"/Users/sahu0957/trisomy_normalization/RNAGRO01132021/data/RNAseqfiles/POWERANALYSIS_1.1_finalsims_a\"+str(trisomy_disp)+\"_p\"+str(trisomy_extrapoiss)+\"_r\"+str(trial_num)+\"_s\"+str(scaling)+\"_n\"+str(replicate_num)+\".csv\"\n",
    "        simulated_df.to_csv(filename, sep=',')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ds_norm",
   "language": "python",
   "name": "ds_norm"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
